### PR TITLE
ENCD-4316 Make generic matrix-rendering component.

### PR DESCRIPTION
**Pull request by forresttanaka**
75% of the point of this ticket is to make a matrix component that doesn't need to be copied and rewritten for every kind of matrix that's coming, as we already did with audit matrix. For new matrix types, the vast majority of the work involves writing a shim to convert the data to a form that the new generic matrix component can render.

The new generic component in this branch is `<DataTable>` which can be thought of as the matrix version of `<SortTable>`, though`<SortTable>` takes data in one object and the table’s shape in another one. `<DataTable>` in contrast takes one object that defines both the data and its shape, mostly because things like column spans are very data dependent and difficult to separate from the shape. I have [a repo for a stand-alone React app to test this matrix code](https://github.com/forresttanaka/matrixcomponent) if you want to play with this component.

I put documentation for the object to pass to `<DataTable>` in datatable.js because of its complexity. The audit matrix JSON format is the same as the experiment matrix, so they both use the same shim with a slightly different root component. `<DataTable>` has no opinion on the data you pass to your shim — only the data that comes out.

The points below probably only need to be read if you have questions that these happen to answer.

* I changed the Encyclopedia BDD test to remove the encyclopedia_version from the query string so that it gets non-zero results. I changed the matrix display to show “No results found” if no results get returned for the given matrix query string, instead of showing an empty matrix with some rendering anomalies (cut-off labels). Also, testing an empty matrix case seemed not that useful to me.
* Some of the color calls use “tint” where they used to use “shade.” The old routines both tinted and shaded, but we never used shading, and the new one only tints, so it seemed better to use the correct terminology. I also removed the shade from many home-page calls as I don't remember why we used that.
* I removed the “color” npm package because, while it still downloaded through npm, going to its repo redirected to a different but similar color package. I found that package just couldn’t do what we wanted, so I added code to datacolors.js to do the job.
* I removed auditmatrix.js completely and just added a small component that uses largely the same code as the experiment matrix except for small changes, removing nearly all the copied code.
* Some of the changes to search-related pages have to do with refactoring out the buttons at the top of the search results. `<SearchControls>` in search.js now renders these. The View buttons (report, matrix, summary, etc.) are now rendered by `<ViewControls>`. This branch also refactored out the “Clear filters” link for the same reasons. The changes to search.js, summary.js, and report.js mostly have to do with using these shared components, and of course matrix.js also shares these.
* Every svg icon now takes an optional CSS style (not class) which I needed to make expander icons either black or white depending on their background. I could also have injected classes, but I already use dynamic styles for similar parts of the matrix, and it seemed a handy ability that can mostly be ignored anyway.
* Why svg chevrons when fontawesome has chevrons? These are flatter, and Ben didn’t want them so tall.
* I found a new version of the [JSDoc style guide](https://github.com/shri/JSDoc-Style-Guide), unrelated to anything from AirBnB or anything like that. I used parts of it for new code in this ticket. I don’t consider this an enforced encoded thing, or even using JSDoc at all. I find it helpful and consistent for myself. I find large parts of that style guide a bit over the top though.

### CODE REVIEW

File: src/encoded/static/components/matrix.js [hunk 1]
```diff
@@ -1,367 +1,589 @@
...
+ *     {
+ *         {array} subCategorySums: Column sums for all subcategory rows
+ *         {number} maxSubCategoryValue: Maximum value in all cells in all subcategory rows
+ *         {number} minSubCategoryValue: Minimum value in all cells in all subcategory rows
+ *     }
+ */
+const analyzeSubCategoryData = (subCategoryData, columnCategoryType) => {
+    const subCategorySums = [];
+    let maxSubCategoryValue;
+    let minSubCategoryValue;
```
Comment (emro) [inline src/encoded/static/components/matrix.js]: Could we just initialize these to be 0 and skip the check for if they are undefined in line 162?

File: src/encoded/static/scss/encoded/modules/_auditmatrix.scss [hunk 2]
```diff
@@ -78,28 +78,28 @@ th.rotate30 {
...
+//     td {
+//         vertical-align: middle;
+//         text-align: center;
+//         padding: 0 !important;
+//         border: solid 1px #f9f9f9;
+//     }
+//     tfoot {
+//         background: transparent;
+//     }
+// }
```
[the remark this answers is not in the document]
Comment (forresttanaka) [reply src/encoded/static/scss/encoded/modules/_auditmatrix.scss]: No! Removed.

File: src/encoded/static/components/matrix.js [hunk 3]
```diff
@@ -1,367 +1,589 @@
...
+            // Right edge of matrix scrolled into view.
+            this.setState({ rightShadingShowing: false });
+        } else if (element.scrollLeft > 0 && !this.state.leftShadingShowing) {
+            // Left edge of matrix scrolled out of view.
+            this.setState({ leftShadingShowing: true });
+        } else if (element.scrollLeft + element.clientWidth < element.scrollWidth && !this.state.rightShadingShowing) {
+            // Right edge of matrix scrolled out of view.
+            this.setState({ rightShadingShowing: true });
+        }
+    }
```
[the remark this answers is not in the document]
Comment (forresttanaka) [reply src/encoded/static/components/matrix.js]: I think we might want to merge that in the future. I saw your function but wanted to see if I could do it without direct DOM manipulation. I’d say that’s a future ticket.

File: src/encoded/static/scss/encoded/modules/_matrix.scss [hunk 4]
```diff
@@ -1,113 +1,360 @@
...
     }
-    tfoot {
-        background: transparent;
+}
+
+
+// note the position of the inner span isn't calculated correctly yet
+$rotated-th-height: 130px;
+$rotated-th-width: 20px;
+$rotated-th-angle: 30deg;
```
[the remark this answers is not in the document]
Comment (forresttanaka) [reply src/encoded/static/scss/encoded/modules/_matrix.scss]: No. Removed.

File: package-lock.json [hunk 5]
```diff
@@ -4362,12 +4371,14 @@
...
-          "dev": true
+          "dev": true,
+          "optional": true
         },
         "yallist": {
           "version": "3.0.2",
           "bundled": true,
-          "dev": true
+          "dev": true,
+          "optional": true
```
[the remark this answers is not in the document]
Comment (forresttanaka) [reply package-lock.json]: Re-ran `npm i` and this time it removed all the optionals. I have no idea what's going on, but committed this updated one with the optionals removed. Recent versions of npm seem a little inscrutable.

File: src/encoded/static/components/datatable.js [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+};
+
+
+/**
+ * Display a table of data with customizable styles per cell and per row, and with column spans.
+ */
+const DataTable = ({ tableData }) => {
+    let colNumber;
+    let cellContent;
+    let maxWidth;
```
[the remark this answers is not in the document]
Comment (forresttanaka) [reply src/encoded/static/components/datatable.js]: I did it this way because tableDataMaxWidth isn’t necessarily inexpensive if the table has a lot of columns, and we only need to call this if tableData has a 0 colSpan somewhere in it. If tableData has no colSpan:0 element, we never call tableDataMaxWidth.

File: src/encoded/static/components/matrix.js (hunk 1 above)
Comment (forresttanaka) [reply src/encoded/static/components/matrix.js]: Almost. With minSubCategoryValue set to 0, the result for that is always zero. But I just learned Javascript has a defined MAXINT referred to as `Number​.MAX_VALUE` so I initialize minSubCategoryValue to that, with maxSubCategoryValue set to 0.

File: src/encoded/static/scss/encoded/modules/_facet.scss [hunk 7]
```diff
@@ -59,34 +60,20 @@
...
-    padding-bottom: 25px;
+    padding-bottom: 10px;
 }
 
-.col-sm-5 {
-    .clear-filters-control {
+.clear-filters-control {
+    position: relative;
+    height: 1.1rem;
+
```
[the remark this answers is not in the document]
Comment (forresttanaka) [reply src/encoded/static/scss/encoded/modules/_facet.scss]: I added it. Yeah, that looked kind of tight. It makes the one on the Summary page look a tad far down, but we might want to address that in a future summary-page update.

I also got rid of the following .clear-filters-control-matrix definition because I don’t use it in matrix.js anymore.

File: src/encoded/static/components/matrix.js (hunk 1 above)
Comment (forresttanaka) [reply src/encoded/static/components/matrix.js]: Also looking at this revealed a very subtle bug in that analyzeSubCategoryData wasn't filtering out zero values in the data when it looked for minimum values, so it always returned zero in minSubCategoryValue anyway. I changed it to filter out zero values when checking for minimums. Also, I subtract one from the minimum value before returning it (minimum value is always guaranteed to be ≥ 1 now). Without that, minimum values in the table would always be white, which I think people would complain about. By subtracting from the minimum value, any non-zero values have at least some obvious color in the table.

File: package-lock.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (emro) [reply package-lock.json]: Sounds good!

File: src/encoded/static/components/datatable.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (emro) [reply src/encoded/static/components/datatable.js]: Ah yes, this makes sense.

File: src/encoded/static/components/matrix.js (hunk 1 above)
Comment (emro) [reply src/encoded/static/components/matrix.js]: Glad you caught that! I agree that the non-zero values should not be white. I wonder if it might have been more straightforward to use a value that isn't 0 for full-width rows but I can't really think of a better value.
